csrc/quantization/machete/machete_mm_kernel.cuh

```diff
@@ -0,0 +1,215 @@
+            ArchTag, OperatorClass, TileShape, ClusterShape, EpilogueTileType,
+            ElementAccumulator, ElementAccumulator, ElementC, LayoutC_Transpose,
+            AlignmentC, ElementD, LayoutD_Transpose, AlignmentD,
+            EpilogueSchedule>::CollectiveOp;
+
+    using CollectiveMainloop =
+        typename cutlass::gemm::collective::VLLMCollectiveBuilder<
+            cutlass::gemm::collective::MacheteKernelTag, ArchTag, OperatorClass,
+            BTypeTuple, PrepackedLayoutBB, AlignmentB, ElementA,
+            LayoutA_Transpose, AlignmentA, ElementAccumulator, TileShape,
+            ClusterShape,
+            cutlass::gemm::collective::StageCountAutoCarveout<static_cast<int>(
+                sizeof(typename CollectiveEpilogue::SharedStorage))>,
+            KernelSchedule>::CollectiveOp;
+
+    using GemmKernel = cutlass::gemm::kernel::GemmUniversal<
+        Shape<int, int, int, int>,  // Indicates ProblemShape
+        CollectiveMainloop, CollectiveEpilogue, TileScheduler>;
+    using Gemm = cutlass::gemm::device::GemmUniversalAdapter<GemmKernel>;
+
+    using StrideA = cutlass::detail::TagToStrideA_t<LayoutA>;
+    using StrideB = cutlass::detail::TagToStrideB_t<LayoutB>;
+    using StrideC = typename GemmKernel::StrideC;
+    using StrideD = typename GemmKernel::StrideD;
+    using StrideS = typename CollectiveMainloop::StrideScale;
+
+    using Arguments = typename Gemm::Arguments;
+    using MainloopArguments = typename GemmKernel::MainloopArguments;
+    using EpilogueArguments = typename GemmKernel::EpilogueArguments;
+
+    static Arguments create_arguments(cudaStream_t stream, int M, int N, int K,
+                                      ElementA const* A, ElementB const* B,
+                                      ElementC const* C, ElementD* D,
+                                      ElementScale const* scales,
+                                      ElementZero const* zeros,
+                                      ElementCompute alpha, ElementCompute beta,
+                                      std::optional<int> maybe_group_size) {
+      // if we have zeropoints we need scales
+      static_assert(!with_zeropoints || with_scales);
+      // if beta != 0 then we need C
+      TORCH_CHECK(with_C || (!with_C && beta == 0));
+      // if with_scales, we need a scales pointer
+      TORCH_CHECK(with_scales || !scales);
+      // if with_zeropoints, we need a zeros pointer
+      TORCH_CHECK(with_zeropoints || !zeros);
+
+      static int constexpr L = 1;
+      int const group_size = maybe_group_size.value_or(K);
+      int const scale_k = (K + group_size - 1) / group_size;
+
+      // not stride_B is unused
+      auto stride_A = make_cute_stride(StrideA{}, N, K, L);
+      auto stride_B = make_cute_stride(StrideB{}, M, K, L);
+      auto stride_C = make_cute_stride(StrideC{}, N, M, L);
+      auto stride_D = make_cute_stride(StrideD{}, N, M, L);
+      auto stride_S = make_cute_stride(StrideS{}, N, scale_k, L);
```

Comment:
There's a particular limitation of `make_cute_stride` and its usage here that I think we should try to avoid.

Setting strides based on the shapes of the tensors prevents us from passing subtensors into kernels. The machete kernels would fail a unit test like this: https://github.com/vllm-project/vllm/blob/07ab160741a486bbef23efbf26aaf2ea8a785ae1/tests/kernels/test_cutlass.py#L342-L345 This limitation could, for example, prevent us from optimizing away slice or concat operations with future torch.compile optimizations.

A bunch of kernels in vLLM do have this limitation, but if we're making nice utils to set up cute strides, we should take the strides into account like we do here:
https://github.com/vllm-project/vllm/blob/07ab160741a486bbef23efbf26aaf2ea8a785ae1/csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu#L423-L433

Reply:
cool made a `make_cute_layout` util that handles this and updated the launcher to use that.

good call :+1: